block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -133,57 +135,109 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
   def insert(
       block: BlockMessage,
       invalid: Boolean
-  ): F[BlockDagRepresentation[F]] =
-    lock.withPermit(
+  ): F[BlockDagRepresentation[F]] = {
+    import cats.instances.list._
+    import coop.rchain.catscontrib.Catscontrib.ToBooleanF
+
+    val logAlreadyStored =
+      Log[F].warn(s"Block ${PrettyPrinter.buildString(block, short = true)} is already stored.")
+
+    def doInsert(blockMetadata: BlockMetadata): F[Unit] = {
+      val updateLatestMessages = {
+        val sendersNewLM  = (block.sender, block.blockHash)
+        val senderIsEmpty = block.sender == ByteString.EMPTY // Genesis has empty sender
+        val logEmptySender =
+          Log[F].warn(s"Block ${Base16.encode(block.blockHash.toByteArray)} sender is empty")
+
+        /** For newly bonded validators there are several options. In general, we should be as lazy as possible.
+          * And probably should add LM only when newly bonded validator creates a block.
+          * But for now we set the block where validator has been bonded as its first Latest Message.
+          * This logic is subject to some research. **/
+        // If validator is bonded but does not have self-justification in block, this is newly bonded validator
+        val newlyBondedSet = bonds(block)
+          .map(_.validator)
+          .toSet
+          .diff(block.justifications.map(_.validator).toSet)
+
+        for {
+          // This filter is required to enable adding blocks backward from higher height to lower
+          newlyBondedUnseen <- newlyBondedSet.toList.filterA(
+                                lm => latestMessagesIndex.contains(lm).not
+                              )
+          newlyBondedLMs = newlyBondedUnseen.map(v => (v, block.blockHash)).toSet
+          // As genesis has empty sender, in the very first block sender will be in newlyBondedUnseen
+          // (no self justification). In this case LM should be updated. It interferes with protection from
+          // regressive seq number for LM, so needs some additional logic to check.
+          sendersLmIsGenesis <- for {
+                                 sendersCurLM <- latestMessagesIndex.get(block.sender)
+                                 r <- sendersCurLM match {
+                                       case None => false.pure[F]
+                                       case Some(lm) =>
+                                         blockMetadataIndex.getUnsafe(lm).map(_.seqNum).map(_ == 0)
+                                     }
+                               } yield r
+
+          // Do not create LM for empty sender, which is a now valid sender for genesis
+          lms <- if (senderIsEmpty || (newlyBondedSet.contains(block.sender) && !sendersLmIsGenesis))
```

Comment:
Why is necessary to have `sendersLmIsGenesis`?

Reply:
Does not the comment above answer?
https://github.com/rchain/rchain/blob/d7b308bdfe4ef36b89734a70c87e5e3da53dc53e/block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala#L168-L179

Reply:
LM for validator is created when processing genesis. But the very first block does not have self justification, so won't try to update lates message assuming it is regressive. Imagine adding block from the top to the bottom - sender has LM all the time and in the end meet block with no self-justification. In general, we should not update LM in this case. But here is exactly the same situation, except we should update LM.
If we would create LM when block is created by sender, not newly bonded - we wont need to have this workaround. Or if genesis can have a set of senders.